From dc78fd3a7bcf06627dc84446cf5fb1004dcbd660 Mon Sep 17 00:00:00 2001
From: Duy Truong <duytruong@google.com>
Date: Mon, 4 May 2026 16:06:25 -0700
Subject: [PATCH] Prevent SQL injection in SelectionBuilder

Use SQLiteTokenizer.OPTION_CHECK_BRACKETS to ensure that selection
clauses added to SelectionBuilder do not have unbalanced brackets.
This prevents callers from breaking out of the parentheses used
when concatenating clauses.

Bug: 393377017
Test: atest SelectionBuilderTest
Flag: EXEMPT BUGFIX

Autogenerated by MendIt (go/androidmendit)
Cherrypick-From: https://googleplex-android-review.googlesource.com/q/commit:d7a97812d4b58d033b81d8d2ebcb24d1f8a5edd8
Merged-In: If178bd2ab556147d7e86771e6157c531499d8ff5
Change-Id: If178bd2ab556147d7e86771e6157c531499d8ff5
---
 .../contacts/util/SelectionBuilder.java         |  2 ++
 .../contacts/util/SelectionBuilderTest.java     | 17 +++++++++++++++++
 2 files changed, 19 insertions(+)

diff --git a/src/com/android/providers/contacts/util/SelectionBuilder.java b/src/com/android/providers/contacts/util/SelectionBuilder.java
index 14499e8b..e21e2e29 100644
--- a/src/com/android/providers/contacts/util/SelectionBuilder.java
+++ b/src/com/android/providers/contacts/util/SelectionBuilder.java
@@ -16,6 +16,7 @@
 
 package com.android.providers.contacts.util;
 
+import android.database.sqlite.SQLiteTokenizer;
 import android.text.TextUtils;
 
 import java.util.ArrayList;
@@ -44,6 +45,7 @@ public class SelectionBuilder {
      */
     public SelectionBuilder addClause(String clause) {
         if (!TextUtils.isEmpty(clause)) {
+            SQLiteTokenizer.tokenize(clause, SQLiteTokenizer.OPTION_CHECK_BRACKETS, null);
             mWhereClauses.add(clause);
         }
         return this;
diff --git a/tests/src/com/android/providers/contacts/util/SelectionBuilderTest.java b/tests/src/com/android/providers/contacts/util/SelectionBuilderTest.java
index 06f7e297..5d96c902 100644
--- a/tests/src/com/android/providers/contacts/util/SelectionBuilderTest.java
+++ b/tests/src/com/android/providers/contacts/util/SelectionBuilderTest.java
@@ -54,4 +54,21 @@ public class SelectionBuilderTest extends FixedAndroidTestCase {
         assertEquals("(A)", new SelectionBuilder("").addClause("A").build());
         assertEquals("(A) AND (B)", new SelectionBuilder("A").addClause("B").build());
     }
+
+    public void testInjection() {
+        try {
+            new SelectionBuilder("1=1)) OR ((1=1");
+            fail("Should have thrown");
+        } catch (IllegalArgumentException expected) {
+            // expected
+        }
+
+        try {
+            SelectionBuilder sb = new SelectionBuilder("A");
+            sb.addClause("1=1)) OR ((1=1");
+            fail("Should have thrown");
+        } catch (IllegalArgumentException expected) {
+            // expected
+        }
+    }
 }
-- 
2.55.0.897.gb25b4bd76c-goog

